### PR TITLE
ARROW-12837: [C++] Do not crash when printing invalid arrays

### PR DESCRIPTION
**Pull request by amol-**

### CODE REVIEW

**Comment by pitrou** [comment]
I'm not opposed to this a priori, but I would find it better if the test suite avoided creating invalid arrays (except where specifically desired, e.g. to test validation).

**Comment by amol-** [comment]
> I'm not opposed to this a priori, but I would find it better if the test suite avoided creating invalid arrays (except where specifically desired, e.g. to test validation).

I wasn't much concerned about the test suite, which I agree can create valid arrays (creating null arrays smells like a form of premature optimization). I think that my focus was more toward C++ users ending up with a `nullptr` returned from somewhere else. In such case it makes your life a little easier if you just get printed out that the array is invalid instead of having to attach `gdb` to understand why you are facing a crash

**Comment by pitrou** [comment]
As I said, why not. But any work trying to enhance the validity check here will end up reproducing what's already done by `Validate` and `ValidateFull`.

**Comment by amol-** [comment]
> As I said, why not. But any work trying to enhance the validity check here will end up reproducing what's already done by `Validate` and `ValidateFull`.

Thanks for pointing out `Validate`. I tried to replace the usage of my custom method with a call to `Validate` and it passes my test, so it's already verifying for the condition I was trying to catch. Locally I had 9 failures when running all tests but were unrelated ( most of them about `PARQUET_TEST_DATA` not being set), so I hope that we can just reuse validate without any side effect.

**Comment by amol-** [comment]
@pitrou I think this is ready for a second review, what do you think?

**Comment by pitrou** [comment]
@bkietz Would you have an opinion on this?

**Review by bkietz — CHANGES_REQUESTED** [review]
I think this is acceptable, since it currently only costs a call to Array::Validate which is relatively cheap. My only worry would be that users might consider (for example) an Int32Array with null data Buffer is de facto acceptable to construct since ToString() doesn't crash. I think the message is explicit enough that if someone opens a JIRA like "segfault with my array but  ToString()'s fine" we can explain that ToString() is more permissive than other array methods

File: cpp/src/arrow/pretty_print.cc [hunk 1]
```diff
@@ -325,6 +325,12 @@ class ArrayPrinter : public PrettyPrinter {
                   std::is_base_of<FixedSizeListArray, T>::value,
               Status>
   Visit(const T& array) {
+    Status validArrayState = array.Validate();
+    if (!validArrayState.ok()) {
+      (*sink_) << "<InvalidArray: " << validArrayState.message() << ">";
```
[the remark this answers is not in the document]
Comment (amol-) [reply cpp/src/arrow/pretty_print.cc]: Applied your suggested change! I should have used underscores, 

Even thought, a bit more talkative variable name would have helped ;)
Even if I get that `st` is a `Status`, it doesn't tell me much about what status it is and I have to navigate to the line of code that returned it to be able to guess, but in this case it's a 3 lines of code block, so I guess that having to look at the top of it to get more context probably doesn't matter too much.

**Review by bkietz — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**